### ISSUE TITLE
Area2D registers collision of CharacterBody2D with disabled collision shape on scene load

### ISSUE THREAD

**Issue by MarcoN96** [godotengine/godot#88592]
### Tested versions

Reproduceable in:
- Godot Engine v4.3.dev3.mono.official [36e943b6b]
- Godot Engine v4.2.1.stable.mono.official [b09f793f5]
- Godot Engine v4.1.3.stable.mono.official [f06b6836a]
- Godot Engine v4.0.4.stable.mono.official [fc0b241c9]

### System information

Godot v4.2.1.stable.mono - Windows 10.0.19045 - Vulkan (Forward+) - dedicated NVIDIA GeForce GTX 1080 (NVIDIA; 31.0.15.5123) - AMD Ryzen 9 5900X 12-Core Processor (24 Threads)

### Issue description

A scene is setup with 4 Area2Ds with a size of 512x512 placed around the origin.
In the main scene this scene is instantiated and a CharacterBody2D with it's CollisionShape2D initially disabled is also instantiated.
The CharacterBody2D is set to position (256, 256) and then it's CollisionShape2D is enabled with a SetDeferred() call.

Expected behaviour: only one Area2D's OnBodyEntered() function is triggered.
Actual behaviour: all 4 Area2D's OnBodyEntered() function is triggered, then the other 3 Area2D's OnBodyExited() function is triggered.

### Steps to reproduce

Using the MRP:
- open the project in the editor
- either run main.tscn in the dotnet folder or main.tscn in the gdscript folder
- watch the messages in the output window
- press spacebar to reload the scene and recreate the issue

### Minimal reproduction project (MRP)

[CollisionBug.zip](https://github.com/godotengine/godot/files/14343746/CollisionBug.zip)

alternative the repo is also available on [github](https://github.com/MarcoN96/GodotCollisionBug)

**Comment by dalexeev** [comment]
See also:

* #14578
* #61584

**Comment by MarcoN96** [comment]
For anyone stumbling upon this issue in the future:
A workaround I found is adding a flag and timer to the Area2D. as long as the flag is false the collision handling functions don't do any processing. On timeout of the timer the flag is set to true.
A second timer on the CharacterBody2D which ideally is longer then the first one enables the CollisionShape2D. This way only the correct Area2D's OnBodyEntered function is triggered.

Despite a workaround existing I would still consider the current behaviour a bug.

**Comment by AlexOtsuka** [comment]
I have encountered this before, and here are a few things I know or believe about it:

- The whole issue is due to the `CharacterBody2D`'s default `GlobalPosition` being `x: 0, y: 0` when `add_child()` is called. You would expect setting its `GlobalPosition` on the very next line would override this, but the `PhysicsServer2D` has already taken that initial position into account.

- Due to this, setting the `CharacterBody2D`'s `GlobalPosition` right before calling `add_child()` actually causes the collisions to behave as intended without using any workaround.

- If you need to set the `GlobalPosition` after the `add_child()` instruction for whatever reason, **awaiting exactly 2 physics frames** before enabling the `CollisionShape2D` will also result in the proper collision signals. This seems to give the `PhysicsServer2D` enough time to start calculating the collisions from the proper `GlobalPosition`. In fact, a very similar workaround was shared in [this comment](https://github.com/godotengine/godot/issues/14578#issuecomment-351253711), all the way back in Godot 2.1.4 over 6 years ago.

- I ran into the same issue using `get_tree().paused = true` before modifying a `CharacterBody2D`'s `Position`, resulting in its previous position being used for collision calculations after unpausing.

**Comment by keysole** [comment]
It's bugged in 4.2.2 rc1 as well. Setting global position before adding didn't help, but waiting for 2 process frames did. Also, I don't think the default position is (0, 0), it's the position the node had before it was removed, at least it's 100% this way in my case. I also checked what the area thinks the body's position is in _body_entered signal and it actually prints the correct position, but the signal still fires.

**Comment by AlexOtsuka** [comment]
> It's bugged in 4.2.2 rc1 as well. Setting global position before adding didn't help, but waiting for 2 process frames did. Also, I don't think the default position is (0, 0), it's the position the node had before it was removed, at least it's 100% this way in my case. I also checked what the area thinks the body's position is in _body_entered signal and it actually prints the correct position, but the signal still fires.

Yes you're right about the position, it was only (0, 0) specifically in this case because it was instantiating a fresh Node. It is indeed taking the Node's previous position into account.

**Comment by o-luib** [comment]
Hi,
this bug seems close enough to the issue I'm facing with reparenting the collider in a _on_area3D_body_entered method. I'm using v4.4.beta3.

I didn't isolate the issue in a MRP and it's just my second godot project, so take the information with a grain of salt.

**Here's the TLDR of my observations**
1. when the body is teleported out of the overlapping area by reparenting, the body_exited() method is called, but the body is still kept as overlappingBody
2. body_entered() is triggered for all CollisionAreas that have the body as overlapping body when reparenting the body. This even happens when the body got disabled or the bodies collision_layer is set to false
3. after 1. body_entered() is not triggered when the RigidBody enters the CollisionArea physically, likely because it's already a overlappingBody
4. i didn't find a method to clear the overlappingBodies by code

<details>
<summary>Here's the full description</summary>
I'm implementing a dodgeball game with 1 player, 7 bots(CharacterBody with a PickupRange-Area3D having a ConvexPolygonShape3D ) and 1 ball (RigidBody3D)
The bots pickup the ball when it enters the PickupRange-Area3D. During pickup the ball gets disabled and set to a position outside of the PickupRange-Area3D. After that, the ball is enabled again and thrown towards an opposing player.

My Issue: The game will come to a halt eventually, because a bot cannot pick up a ball that it has thrown before.

I unsuccessfully tried: 
- disabling the ball before reparenting
- call_deferred("reparent",...)
- re-position the ball out of the collision area before reparenting
- disabling collision layers on the ball before reparenting

What i did not try yet:
- waiting 2 frames
- disable monitoring before reparenting

So I investigated the RigidBody and Area3D interaction in more detail. Here are some commented logs drilling into the "pickup & throw" flow of the first bot that got in contact with the ball.

```
Ball1 entered Bot2 pickup collision area
Bot2.overlappingBodies=[Ball1:<RigidBody3D#64055413727>]
Bot2 picking up Ball1
disabling Ball1
Ball1 disabled
reparenting Ball1 from Main to Bot2
Ball1 has exited Bot2 pickup area
	### here starts anoter async body_entered callback after reparenting, see summary 1
	### this should not happen at all imo. especially for a disabled body.
	Ball1 entered Bot2 pickup collision area
	Bot2.overlappingBodies=[Ball1:<RigidBody3D#64055413727>]
	Bot2 already picking up Ball1
Ball1 reparented
Bot2 throwing Ball1
reparenting Ball1 from Bot2 to Main
	### same as above, now when putting the ball back "on the field"
	Ball1 has exited Bot2 pickup area
	### The balls position is definitely not inside the Area3D here, but the body entered is still called
	Ball1 entered Bot2 pickup collision area
	Bot2.overlappingBodies=[Ball1:<RigidBody3D#64055413727>]
	Bot2 still hasBall Ball1
```

And the same for the last bot:
```
Ball1 entered Bot1 pickup collision area
Bot1.overlappingBodies=[Ball1:<RigidBody3D#64055413727>]
Bot1 picked up Ball1
disabling Ball1
Ball1 disabled
reparenting Ball1 from Main to Bot1
	Ball1 has exited Bot2 pickup area
	Ball1 has exited Bot4 pickup area
	Ball1 has exited Bot7 pickup area
	Ball1 has exited Bot3 pickup area
	Ball1 has exited Bot5 pickup area
	Ball1 has exited Bot6 pickup area
	Ball1 has exited Bot1 pickup area
	Ball1 entered Bot2 pickup collision area
	Bot2.overlappingBodies=[Ball1:<RigidBody3D#64055413727>]
	Ball1 entered Bot4 pickup collision area
	Bot4.overlappingBodies=[Ball1:<RigidBody3D#64055413727>]
	Ball1 entered Bot7 pickup collision area
	Bot7.overlappingBodies=[Ball1:<RigidBody3D#64055413727>]
	Ball1 entered Bot3 pickup collision area
 	Bot3.overlappingBodies=[Ball1:<RigidBody3D#64055413727>]
	Ball1 entered Bot5 pickup collision area
	Bot5.overlappingBodies=[Ball1:<RigidBody3D#64055413727>]
	Ball1 entered Bot6 pickup collision area
	Bot6.overlappingBodies=[Ball1:<RigidBody3D#64055413727>]
	Ball1 entered Bot1 pickup collision area
	Bot1.overlappingBodies=[Ball1:<RigidBody3D#64055413727>]
	Bot1 already picking up Ball1

Ball1 reparented
Bot1 throwing Ball1
reparenting Ball1 from Bot1 to Main
Ball1 has exited Bot2 pickup area
[ ... same for all other bots]
Ball1 entered Bot1 pickup collision area
Bot1.overlappingBodies=[Ball1:<RigidBody3D#64055413727>]
Bot1 already hasBall Ball1
```

Eventually every Bot has Ball1 as overlappingBody no matter where they are and they can't collide with the ball physically anymore. body_entered() is only triggered after reparenting. 

```
Bot1 has overlappingBodies: [Ball1:<RigidBody3D#64055413727>]
Bot2 has overlappingBodies: [Ball1:<RigidBody3D#64055413727>]
Bot3 has overlappingBodies: [Ball1:<RigidBody3D#64055413727>]
Bot5 has overlappingBodies: [Ball1:<RigidBody3D#64055413727>]
Bot6 has overlappingBodies: [Ball1:<RigidBody3D#64055413727>]
Bot7 has overlappingBodies: [Ball1:<RigidBody3D#64055413727>]
```

</details>
